pkg/registry/core/service/storage/rest.go

```diff
@@ -267,6 +269,22 @@ func (rs *REST) Delete(ctx context.Context, id string, deleteValidation rest.Val
 	return status, true, nil
 }
 
+func (rs *REST) getAfterDeleteFunc() genericregistry.ObjectFunc {
+	return func(obj runtime.Object) error {
+		svc, ok := obj.(*api.Service)
+		if !ok {
+			return fmt.Errorf("unexpected type of deleted object, want service")
+		}
+		rs.releaseAllocatedResources(svc)
+		ctx := genericapirequest.WithNamespace(context.Background(), svc.Namespace)
+		_, _, err := rs.endpoints.Delete(ctx, svc.Name, rest.ValidateAllObjectFunc, &metav1.DeleteOptions{})
```

Comment:
Can you please comment this?  Why are we catching the AfterDelete hook, just to delete the object again?

Reply:
these delete the endpoints, we were deleting the service before, no?

Reply:
Actually, @thockin @liggitt , IIUIC,  it seems that this `endpoints.Delete()` should be removed because it will race with the endpoints controller

https://github.com/kubernetes/kubernetes/blob/4fbf5df52b122ef9f466676792b8afed7157ced5/pkg/controller/endpoint/endpoints_controller.go#L362-L367

another thing is that  this is inconsistent with endpointslices, that is the default since 1.19

Reply:
> it seems that this `endpoints.Delete()` should be removed because it will race with the endpoints controller

Even though I agree it is redundant in theory, I'm hesitant to remove the synchronous delete that has been in place since 1.0.

> another thing is that this is inconsistent with endpointslices, that is the default since 1.19

EndpointSlices is still in beta, I would not change the GA behavior to match the beta behavior